resolver/resolver.go

```diff
@@ -67,147 +71,217 @@ func (r *Resolver) ResolveToLastNode(ctx context.Context, fpath path.Path) (cid.
+		return nil, err
+	}
+
+	c, p, err := path.SplitAbsPath(fpath)
 	if err != nil {
-		evt.Append(logging.LoggableMap{"error": err.Error()})
 		return nil, err
 	}
 
-	log.Debug("resolve dag get")
-	nd, err := r.DAG.Get(ctx, h)
+	// create a selector to traverse all path segments but only match the last
+	pathSelector, err := pathAllSelector(p)
 	if err != nil {
-		evt.Append(logging.LoggableMap{"error": err.Error()})
 		return nil, err
 	}
 
-	return r.ResolveLinks(ctx, nd, parts)
+	nodes, _, _, err := r.resolveNodes(ctx, c, pathSelector)
+	return nodes, err
 }
 
 // ResolveLinks iteratively resolves names by walking the link hierarchy.
-// Every node is fetched from the DAGService, resolving the next name.
+// Every node is fetched from the fetcher, resolving the next name.
 // Returns the list of nodes forming the path, starting with ndd. This list is
 // guaranteed never to be empty.
 //
 // ResolveLinks(nd, []string{"foo", "bar", "baz"})
 // would retrieve "baz" in ("bar" in ("foo" in nd.Links).Links).Links
-func (r *Resolver) ResolveLinks(ctx context.Context, ndd ipld.Node, names []string) ([]ipld.Node, error) {
+func (r *Resolver) ResolveLinks(ctx context.Context, ndd ipldp.Node, names []string) ([]ipldp.Node, error) {
 
 	evt := log.EventBegin(ctx, "resolveLinks", logging.LoggableMap{"names": names})
 	defer evt.Done()
-	result := make([]ipld.Node, 0, len(names)+1)
-	result = append(result, ndd)
-	nd := ndd // dup arg workaround
-
-	// for each of the path components
-	for len(names) > 0 {
-		var cancel context.CancelFunc
-		ctx, cancel = context.WithTimeout(ctx, time.Minute)
-		defer cancel()
-
-		lnk, rest, err := r.ResolveOnce(ctx, r.DAG, nd, names)
-		if err == dag.ErrLinkNotFound {
-			evt.Append(logging.LoggableMap{"error": err.Error()})
-			return result, ErrNoLink{Name: names[0], Node: nd.Cid()}
-		} else if err != nil {
-			evt.Append(logging.LoggableMap{"error": err.Error()})
-			return result, err
+
+	// create a selector to traverse all path segments but only match the last
+	pathSelector, err := pathAllSelector(names)
+	if err != nil {
+		return nil, err
+	}
+
+	// create a new cancellable session
+	ctx, cancel := context.WithTimeout(ctx, time.Minute)
+	defer cancel()
+
+	session := r.FetchConfig.NewSession(ctx)
+
+	// traverse selector
+	nodes := []ipldp.Node{ndd}
+	err = session.NodeMatching(ctx, ndd, pathSelector, func(res fetcher.FetchResult) error {
+		nodes = append(nodes, res.Node)
+		return nil
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	return nodes, err
+}
+
+// Finds nodes matching the selector starting with a cid. Returns the matched nodes, the cid of the block containing
+// the last node, and the depth of the last node within its block (root is depth 0).
+func (r *Resolver) resolveNodes(ctx context.Context, c cid.Cid, sel selector.Selector) ([]ipldp.Node, cid.Cid, int, error) {
+	// create a new cancellable session
+	ctx, cancel := context.WithTimeout(ctx, time.Minute)
```

Comment:
+1, I'd rather keep pushing these up and up and up until they're an application choice.

Reply:
I'm leaving till we here from @Stebalien -- it seems important to preserve compatible behavior at least on a first release, since the timeouts are effectively from the previous code (albeit in a slightly different place)

Reply:
So I think per https://github.com/ipfs/go-path/pull/36#discussion_r675739954, we're leaving timeouts in, for this release?  I'll mark this resolved, then.  (I do hope we manage to raise them out and above this level in the future, though.)